### PR TITLE
Cleaning up/adding new faq to app starters docs

### LINKED ISSUE #66: Port app build/customization FAQs to reference guide
As a user, I'd like to review the steps to customize, build and generate (_with suitable binder_) applications either as a standalone artifact or in bulk.

**Acceptance:**
FAQs from Google doc are formally ported into official reference guide.

### PR DESCRIPTION
**Pull request by sobychacko**
Fixes #66

### CODE REVIEW

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 1]
```diff
@@ -83,15 +85,15 @@ The Docker versions of the applications are available in Docker Hub, at `https:/
...
 ----
 
-will pull the _latest_ Docker image of the _Cassandra sink_ with the Kafka binder.
+will pull the _latest_ Docker image of the _Cassandra sink_ with the Kafka binder that is running with Kafka version 0.10.
 
 ==== Building the artifacts
 
 You can also build the project and generate the artifacts (including the prebuilt applications) on your own.
-This is useful if you want to deploy the artifacts locally, for example for adding a new starter, or if you want to build the entire set of artifacts available in a starter repository with a new binder.
+This is useful if you want to deploy the artifacts locally or patch the starter with some customizations or missing features and build again.
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: or add additional features

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 2]
```diff
@@ -363,22 +366,20 @@ configuration section of the plugin.
 </configuration
 ----
 
-By default, we generate apps for both Kafka and Rabbitmq binders - `spring-cloud-stream-binder-kafka` and
-`spring-cloud-stream-binder-rabbit`. Say, if you have a custom binder you created for some middleware (say JMS),
+By default, we generate apps for both Kafka 09/10 and Rabbitmq binders - `spring-cloud-stream-binder-kafka` and
+`spring-cloud-stream-binder-rabbit`. Say, if you have a custom binder you created for some middleware (say Hazelcast),
 which you need to generate apps for foobar source, you can add that binder to the binders list in the configuration
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: why hazelcast instead of jms?

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 3]
```diff
@@ -492,4 +494,146 @@ Then do this:
 
 This would upload the docker image to the docker hub in your custom repository.
 
+=== General FAQ on Spring Cloud Stream App Starters
+
+In the following sections, you can find a brief faq on various things that we discussed above and a few other infrastructure related topics.
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: uppercase faq

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 4]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+
+
+. What are Spring Cloud Stream Application Starters? +
+Spring Cloud Stream App Starters are individual repositories hosted under the GitHub organization - spring-cloud-stream-app-starters - on github (http://github.com/spring-cloud-stream-app-starters) +
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: This description doesn't seem that good, saying the are repositories.  The description from the project page seems better  to repeat here ```Spring Cloud Stream Application Starters are Spring Boot based Spring Integration applications that provide integration with external systems.```

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 5]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+Project page: http://cloud.spring.io/spring-cloud-stream-app-starters/
+
+
+. What is the parent for stream app starters? +
+The parent for all app starters is `app-starters-build` which is coming from the core project. https://github.com/spring-cloud-stream-app-starters/core
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc:507]: Does the doc show the actual mvn xml snippit of `<parent/>` ?  That would be a useful addition.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 6]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+
+. What are the contents of the core BOM? +
+In addition to the common artifacts in core, the app-starters-core-dependencies BOM also adds dependency management for spring-cloud-dependencies which will include spring-cloud-stream transitively.
+
+. Where is the core BOM used? +
+There are two places where the core BOM is used. It is used to provide compile time dependency management for all the starters.
+This is defined in the `app-starters-build` artfiact. This same BOM is referenced through the maven plugin configuration for the app generation.
+The generated apps thus will include this bom also in their pom.xml files.
+
+. What are the artifacts that the parent artifact (`app-starters-build`) include from the core bom to be part of all the starters by default and there by making them part of the generated apps? +
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: It isn't clear what this question is about.  Is it about what specific spring cloud stream dependencies get pulled in transitively?

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 7]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+Spring-integration bom is imported in the spring-boot-dependencies bom and this is where the default SI dependencies are coming for SCSt app starters.
+
+. Can you summarize all the BOM's that SCSt app starters depend on? +
+All SCSt app starters have access to dependencies defined in the following BOM’s and other dependencies from any other BOM’s these three boms import transitively as in the case of Spring Integration:
+    * app-starters-core-dependencies
+    * spring-cloud-dependencies
+    * spring-boot-dependencies
+
+.  Each app starter has `app-starter-build` as the parent which in turn has `spring-cloud-build` as parent. The above documentation states that the
+generated apps have `spring-boot-starter` as the parent. Where is the mismatch? +
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: Why (not where) is there a mismatch?

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 8]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+There is no mismatch per se, but a slight subtlety. As the question frames, each app starter has access to artifacts managed all the way through `spring-cloud-build` at compile time.
+However, this is not the case for the generated apps at runtime. Their parent is `spring-boot-starter` that imports `spring-boot-dependencies` bom that includes a majority of the components that these apps need.
+The question then is what if the starter used an artifact managed through the `app-starter-build` or `spring-cloud-build`. Remember that the
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: The following two sentences are confusing.  Isn't the answer that generated apps are managed by boot.  The app responsible for generation the app uses a different dependency.  This is so generated apps are consistent with the dependency management used in start.spring.io and start-scs.cfapps.io

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 9]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+`core-app-starter-dependencies` bom is also included in the generated apps. This bom also imports the `spring-cloud-dependencies` bom. This way, at runtime
+the apps have access to everything that it used at compile time.
+
+.  Why is there an app starter specific BOM in each app starer repositories? For example, `time-app-dependencies`. +
+This is an important BOM. At runtime, the generated apps get the versions used in their dependencies through a BOM that is manaing the dependencies. Since all the boms
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: typo managing, not manaing

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 10]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+This bom is provided to the app generator maven plugin in addition to the core bom. This app specific bom has `spring-cloud-dependencies-parent` as parent.
+
+.  How do I create a new app starter project? +
+If you have a general purpose starter that can be provided as an of of the box app, create an issue for that in https://github.com/spring-cloud-stream-app-starters/app-starters-release[app-starters-release].
+If there is a consensus, then a repository can be created in the `spring-cloud-stream-app-starters` organization where you can start contributing the starters and other components.
+
+.  I created a new starter according to the guidelines above, now how do I generate binder specific apps for the new starters? +
+By default, the app-starters-build in core is configured with the common configuration needed for the app generator maven plugin. It is configured for generating apps for kafka-09, kafka-10 and rabbitmq binders.
+In your starter you already have the configuration specified for the plugin from the parent. Override the names accordingly. Refer to an exsiting starter for guidelines.
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: A link to the names in a config file would be useful.  Typo on `existing`.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 11]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+If your app is straightforward, all you have to change is the name of the apps in the <generatedApps> section, and the app-starter specific BOM maven coordinates.
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: What do you mean `if your app is straightforward`?  What does changing the name of the apps do wrt to the binder.  This sentence doesn't seem related to the question.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 12]
```diff
@@ -492,4 +494,146 @@ Then do this:
...
+    <dependency>
+        <groupId>org.springframework.integration</groupId>
+        <artifactId>spring-integration-java-dsl</artifactId>
+        <version>${spring-integration-java-dsl.version}</version>
+    </dependency>
+
+.  How do I override the boot version used in a particular app? +
+When you generate the app, override the boot version as follows.
+
+  ./mvnw clean install -PgenerateApps -DbootVersion=<boot version to override>
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc:644]: Can you give an example showing a specific version, one would specify 1.5.2.RELEASE or 2.0.0.BUILD-SNAPSHOT ?

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 2 above)
Comment (sobychacko) [reply spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: i just randomly picked that up since we already have a jms binder in the works.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 6 above)
Comment (sobychacko) [reply spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: Yes, i will make the question simpler.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 8 above)
Comment (sobychacko) [reply spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: all the dependencies coming from `spring-boot-dependencies` are good here. I was talking about the dependencies specified only at the spring-cloud-dependencies` bom or our specific `core-app-starter` bom. I will try to clarify this in the answer.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 11 above)
Comment (sobychacko) [reply spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: cleaned it up.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 13]
```diff
@@ -367,7 +367,7 @@ configuration section of the plugin.
 ----
 
 By default, we generate apps for both Kafka 09/10 and Rabbitmq binders - `spring-cloud-stream-binder-kafka` and
-`spring-cloud-stream-binder-rabbit`. Say, if you have a custom binder you created for some middleware (say Hazelcast),
+`spring-cloud-stream-binder-rabbit`. Say, if you have a custom binder you created for some middleware (say JMS),
 which you need to generate apps for foobar source, you can add that binder to the binders list in the configuration
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: I think there were a few other spots that had hazelcast...

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 14]
```diff
@@ -494,18 +494,26 @@ Then do this:
...
 
 
 . What is the parent for stream app starters? +
 The parent for all app starters is `app-starters-build` which is coming from the core project. https://github.com/spring-cloud-stream-app-starters/core
+For example:
+
+    <parent>
+        <groupId>org.springframework.cloud.stream.app</groupId>
+        <artifactId>app-starters-build</artifactId>
+        <version>1.2.0.BUILD-SNAPSHOT</version>
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: I think it maybe possible to parameterize this as an asciidoctor variable so the version comes from properties in the maven plugin.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 15]
```diff
@@ -519,7 +527,7 @@ There are two places where the core BOM is used. It is used to provide compile t
 This is defined in the `app-starters-build` artfiact. This same BOM is referenced through the maven plugin configuration for the app generation.
 The generated apps thus will include this bom also in their pom.xml files.
 
-. What are the artifacts that the parent artifact (`app-starters-build`) include from the core bom to be part of all the starters by default and there by making them part of the generated apps? +
+. What spring cloud stream artifacts that the parent artifact (`app-starters-build`) include? +
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: that the parent ->does the parent

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 16]
```diff
@@ -542,15 +550,14 @@ All SCSt app starters have access to dependencies defined in the following BOM
     * spring-boot-dependencies
 
 .  Each app starter has `app-starter-build` as the parent which in turn has `spring-cloud-build` as parent. The above documentation states that the
-generated apps have `spring-boot-starter` as the parent. Where is the mismatch? +
+generated apps have `spring-boot-starter` as the parent. Why is the mismatch? +
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: Why is the mismatch -> Why the mismatch

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 17]
```diff
@@ -542,15 +550,14 @@ All SCSt app starters have access to dependencies defined in the following BOM
...
 There is no mismatch per se, but a slight subtlety. As the question frames, each app starter has access to artifacts managed all the way through `spring-cloud-build` at compile time.
-However, this is not the case for the generated apps at runtime. Their parent is `spring-boot-starter` that imports `spring-boot-dependencies` bom that includes a majority of the components that these apps need.
-The question then is what if the starter used an artifact managed through the `app-starter-build` or `spring-cloud-build`. Remember that the
-`core-app-starter-dependencies` bom is also included in the generated apps. This bom also imports the `spring-cloud-dependencies` bom. This way, at runtime
-the apps have access to everything that it used at compile time.
+However, this is not the case for the generated apps at runtime. Generated apps are managed by boot. Their parent is `spring-boot-starter` that imports `spring-boot-dependencies` bom that includes a majority of the components that these apps need.
+The question then is what if the starter used an artifact managed through the `app-starter-build` or `spring-cloud-build`.
```
Comment (markpollack) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: I still don't get line 556... maybe it can be omitted?

**Comment by markpollack** [comment]
squashed and merged as 987fecb69a3fbc89d41005ecc73d0e7bfcbacee0
